clang/runtime/dpct-rt/include/device.hpp.inc

```diff
@@ -1060,6 +1054,26 @@ static inline sycl::queue &get_default_queue() {
 }
 // DPCT_LABEL_END
 
+// DPCT_LABEL_BEGIN|typedef_queue_ptr|dpct
+// DPCT_DEPENDENCY_BEGIN
+// Device|get_default_queue
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+
+class queue_ptr {
+  sycl::queue* _q;
+public:
+  queue_ptr(sycl::queue* _p)
+    : _q(_p ? _p : &get_default_queue()) {}
+  queue_ptr()
+    : queue_ptr(0) {}
+  operator sycl::queue*  () const { return _q; }
```

Comment:
Why we need this two conversions?

Reply:
Two reasons: (1) for conversion to `sycl::queue*`, this allows us to use operations of a pointer. We can dereference a queue_ptr, use subscripting, add/subtract offset, etc. For conversion to `sycl::queue*&`, this allow use of `+=`, `-=`, `++`, etc. (2) End-to-end test caught the following issue: in other helper header files, there are API that use `sycl::queue*` instead of `queue_ptr`. Specifically `dpct::device_ext::destroy_queue` takes as parameter a reference to a `sycl::queue*`, and one test passed a `queue_ptr` to this. This conversion allows a `queue_ptr` to bind to a this.

Reply:
For reason 1:
`cuaStream_t`/`CUstream_t` is a pointer of an opaque class, which can not be dereferenced. Therefore, `queue_ptr` doesn't need to support such operators.
For reason 2:
I think we should replace `sycl::queue*` with `queue_ptr` in helper functions for consistency.

Reply:
If `queue_ptr` is not dereferenceable, then there is no way to pass the `sycl::queue` is points to by reference to other functions. There are some helper functions that we take as argument a reference to `sycl::queue` - examples include `dpct::dpct_memcpy` and the BLAS utils.

Reply:
I think all the helper function interface should use `queue_ptr` instead of `sycl::queue`. Considering backward compatibility, we also need to add `operator*` for `queue_ptr` which returns `queue_ptr&`.